pkg/eden/vbox.go

```diff
@@ -0,0 +1,251 @@
+			guestPort += 10
+			commandArgsString = fmt.Sprintf("natnetwork  modify --netname %s --port-forward-4 :tcp:[]:%d:[%s]:%d",
+											"natnet1", hostPort, eth1IP.String(), guestPort)
+			if err = utils.RunCommandWithLogAndWait("VBoxManage", defaults.DefaultLogLevelToPrint, strings.Fields(commandArgsString)...); err != nil {
+				log.Fatalf("VBoxManage error for command %s %s", commandArgsString, err)
+			}
+		}
+
+		commandArgsString = fmt.Sprintf("startvm  %s", vmName)
+		if err = utils.RunCommandWithLogAndWait("VBoxManage", defaults.DefaultLogLevelToPrint, strings.Fields(commandArgsString)...); err != nil {
+			log.Fatalf("VBoxManage error for command %s %s", commandArgsString, err)
+		}
+	} else {
+		scanner := bufio.NewScanner(bytes.NewReader([]byte(out)))
+		for scanner.Scan() {
+			line := strings.TrimSpace(scanner.Text())
+			if strings.Contains(line, "VMState=\"poweroff\"") {
+				poweroff = true
+				break
+			}
+		}
+
+		networkFound := false
+		output := ""
+		var err error
+		if output, _, err = utils.RunCommandAndWait("VBoxManage", strings.Fields(fmt.Sprintf("natnetwork list  %s", "natnet1"))...); err != nil {
+			log.Fatalf("VBoxManage error for command natnetwork list %s", err)
+		}
+		scanner = bufio.NewScanner(bytes.NewReader([]byte(output)))
+		for scanner.Scan() {
+			line := strings.TrimSpace(scanner.Text())
+			if strings.HasPrefix(line, "Enabled") {
+				enabled := strings.Split(line, ":")[1]
+				enabled = strings.TrimSpace(enabled)
+				if enabled == "Yes" {
+					networkFound = true
+				}
+			}
+		}
+		if !networkFound {
+			commandArgsString := fmt.Sprintf("natnetwork add --netname %s --network %s --enable --dhcp on",
+			"natnet1", "10.0.2.0/24")
+			if err = utils.RunCommandWithLogAndWait("VBoxManage", defaults.DefaultLogLevelToPrint, strings.Fields(commandArgsString)...); err != nil {
+				log.Fatalf("VBoxManage error for command %s %s", commandArgsString, err)
+			}
+			commandArgsString = fmt.Sprintf("natnetwork start --netname %s", "natnet1")
+			if err = utils.RunCommandWithLogAndWait("VBoxManage", defaults.DefaultLogLevelToPrint, strings.Fields(commandArgsString)...); err != nil {
+				log.Errorf("VBoxManage error for command %s %s", commandArgsString, err)
+			}
+
+			commandArgsString = fmt.Sprintf("modifyvm %s --nic1 natnetwork --nat-network1 %s --cableconnected1 on",
+			vmName, "natnet1")
+			if err = utils.RunCommandWithLogAndWait("VBoxManage", defaults.DefaultLogLevelToPrint, strings.Fields(commandArgsString)...); err != nil {
+				log.Fatalf("VBoxManage error for command %s %s", commandArgsString, err)
+			}
+
+			commandArgsString = fmt.Sprintf("modifyvm %s --nic2 natnetwork --nat-network2 %s --cableconnected2 on",
+			vmName, "natnet1")
+			if err = utils.RunCommandWithLogAndWait("VBoxManage", defaults.DefaultLogLevelToPrint, strings.Fields(commandArgsString)...); err != nil {
+				log.Fatalf("VBoxManage error for command %s %s", commandArgsString, err)
+			}
+		}
+
+		eth0IP := ipMap["eth0"]
```

Comment:
Can we have fallback variant here(with empty ipMap)? In your variant, we expect ipMap filled, but it is true only after EVE onboard.

Reply:
It's not clear to me. Can you please elaborate?

Reply:
You add `getUplinkPortIPMap()` function, which will return non-empty map only in case of onboarded EVE. But in the beginning (not onboarded, just started) it will be empty. So, we should care about it.

Reply:
Yes, you are right. We'll need reboot for the port maps to take affect. Before device onboard, the rules will have the IP as empty string. This should not adversely affect us because we do not run tests before the device in onboarded. After the device is onboarded and rebooted, the port map rules will have valid IP addresses. We already have a reboot in the beginning of tests/workflow that helps setup the port maps properly in this case.
With virtual box, unfortunately we cannot just tell the destination port for packet while configuring host port map rules. It will require the destination IP address for configuring port maps. It's easier with qemu but a bit of inconvenient with vbox

Reply:
Sure, but Eden is not only about the workflow.  Ok, if you think, it is difficult to workaround empty ipMap here, we can think about it in next PRs.

Reply:
My point was that it did not make any difference if we handled the empty ip map or not. It would still require a reboot before the port maps can take effect. Do you have something that proves contrary?